test/examples/ucp_hello_world.c

```diff
@@ -1,17 +1,487 @@
+{
+	struct ucx_context *ctx = (struct ucx_context *) request;
+	ctx->completed = 0;
+}
+
+static int run_server()
+{
+	struct sockaddr_in inaddr;
+	int lsock = -1;
+	int dsock = -1;
+	int optval = 1;
+	int ret;
+
+	lsock = socket(AF_INET, SOCK_STREAM, 0);
+	if (lsock < 0) {
+		fprintf(stderr, "server socket() failed\n");
+		goto err;
+	}
+
+        optval = 1;
+        ret = setsockopt(lsock, SOL_SOCKET, SO_REUSEADDR, &optval, sizeof(optval));
+        if (ret < 0) {
+		fprintf(stderr, "server setsockopt() failed\n");
+		goto err_sock;
+        }
+
+        inaddr.sin_family      = AF_INET;
+        inaddr.sin_port        = htons(server_port);
+        inaddr.sin_addr.s_addr = INADDR_ANY;
+        memset(inaddr.sin_zero, 0, sizeof(inaddr.sin_zero));
+        ret = bind(lsock, (struct sockaddr*)&inaddr, sizeof(inaddr));
+        if (ret < 0) {
+		fprintf(stderr, "server bind() failed\n");
+		goto err_sock;
+	}
+
+	ret = listen(lsock, 0);
+        if (ret < 0) {
+		fprintf(stderr, "server listen() failed\n");
+		goto err_sock;
+        }
+
+        printf("Waiting for connection...\n");
+
+        /* Accept next connection */
+        dsock = accept(lsock, NULL, NULL);
+        if (dsock < 0) {
+		fprintf(stderr, "server accept() failed\n");
+		goto err_sock;
+        }
+
+	close(lsock);
+
+	return dsock;
+
+err_sock:
+	close(lsock);
+err:
+	return -1;
+}
+
+static int run_client(const char *server)
+{
+	struct sockaddr_in conn_addr;
+	struct hostent *he;
+	int connfd;
+	int ret;
+
+	connfd = socket(AF_INET, SOCK_STREAM, 0);
+	if (connfd < 0) {
+		fprintf(stderr, "socket() failed: %m\n");
+		return -1;
+	}
+
+	he = gethostbyname(server);
+	if (he == NULL || he->h_addr_list == NULL) {
+		fprintf(stderr, "host %s not found: %s\n", server, hstrerror(h_errno));
+		goto err_conn;
+	}
+
+	conn_addr.sin_family = he->h_addrtype;
+	conn_addr.sin_port = htons(server_port);
+
+	memcpy(&conn_addr.sin_addr, he->h_addr_list[0], he->h_length);
+	memset(conn_addr.sin_zero, 0, sizeof(conn_addr.sin_zero));
+
+	ret = connect(connfd, (struct sockaddr*)&conn_addr, sizeof(conn_addr));
+	if (ret < 0) {
+		fprintf(stderr, "connect() failed: %m\n");
+		goto err_conn;
+	}
+
+	return connfd;
+
+err_conn:
+	close(connfd);
+
+	return -1;
+}
+
+static void send_handle(void *request, ucs_status_t status)
+{
+	struct ucx_context *context = (struct ucx_context *) request;
+
+	context->completed = 1;
+
+	printf("[0x%x] send handler called with status %d\n",
+	       (unsigned int)pthread_self(), status);
+}
+
+static void recv_handle(void *request, ucs_status_t status,
+			       ucp_tag_recv_info_t *info)
+{
+	struct ucx_context *context = (struct ucx_context *) request;
+
+	context->completed = 1;
+
+	printf("[0x%x] receive handler called with status %d (length %u)\n",
+	       (unsigned int)pthread_self(), status, info->length);
+}
+
+static void wait(ucp_worker_h *ucp_worker, struct ucx_context *context) {
+	while (context->completed == 0)
```

Comment:
correct.
The reason is that I want to demonstrate usage of user defined request context.